Fixed upload-file tmp file space leak

curtisyallop opened the pull request:
Fix version 3 ported to 2.6.0 branch

Cherry picked from:
ckan@c96c216

It is basically the exact same fix only I applied it manually after
getting merge conflicts.

Note that CloseWSGIInputMiddleware must go right after
PluginImplementations plugin.make_middleware loop. The location in the
pylons app pipeline is critical. In the cherry-picked original, it is
right before RootPathMiddleware. But not in this version for 2.6.0. This
RootPathMiddleware line is in a different place in 2.6.0.

Version 3 brings in some improvements changes from ckan. Form vars
named other than "upload" have the leak fixed also. Too-large uploads
do not have the leak when they error. os.rename is moved to a safer
location. Some refactoring was done. Version 2 still essentially fixes
the problem.

Tested on dev-local.

To see the leak:
watch -n .2 'df -h; echo; sudo lsof -c python -c httpd -c nginx | grep
/tmp'

(lsof views deleted "Stale file handle" files which ckan holds on to but
are not visible from "ls" or "find" - this shows the leak)

Original card for 2.3.5
https://trello.com/c/oP4frPB6/200-very-large-files-cannot-be-added-to-open-government-portal

Fix version 2 card: Fix: upload dies - small file uploads and create-links fail
https://trello.com/c/mCTGrKdu/517-fix-large-file-upload-errors?menu=filter&filter=@curtisyallop1

Fix version 2 on 2.3.5 code:
eb4ff13

Fix version 3 ported to ckan 2.6.0 card:
https://trello.com/c/1tDpPqip/449-make-ckan-260-version-of-upload-file-tmp-file-space-leak-fix

neilmayhew commented:
Could you bring this one in line with #3? At least the `try ... finally` and maybe the `hasattr()` as well.

Commit:
Fixed upload-file tmp file space leak

Fix version 3 ported to 2.6.0 branch

Cherry picked from:
https://github.com/ckan/ckan/pull/3756/commits/c96c2163cfe4be9699ea9d804b80f7ed5229f24e

It is basically the exact same fix only I applied it manually after
getting merge conflicts.

Note that CloseWSGIInputMiddleware must go right after
PluginImplementations plugin.make_middleware loop. The location in the
pylons app pipeline is critical. In the cherry-picked original, it is
right before RootPathMiddleware. But not in this version for 2.6.0. This
RootPathMiddleware line is in a different place in 2.6.0.

Version 3 brings in some improvements changes from ckan. Form vars
named other than "upload" have the leak fixed also. Too-large uploads
do not have the leak when they error. os.rename is moved to a safer
location. Some refactoring was done. Version 2 still essentially fixes
the problem.

Tested on dev-local.

To see the leak:
watch -n .2 'df -h; echo; sudo lsof -c python -c httpd -c nginx | grep
/tmp'

(lsof views deleted "Stale file handle" files which ckan holds on to but
are not visible from "ls" or "find" - this shows the leak)

Original card for 2.3.5
https://trello.com/c/oP4frPB6/200-very-large-files-cannot-be-added-to-open-government-portal

Fix version 2 card: Fix: upload dies - small file uploads and create-links fail
https://trello.com/c/mCTGrKdu/517-fix-large-file-upload-errors?menu=filter&filter=@curtisyallop1

Fix version 2 on 2.3.5 code:
https://github.com/abgov/ckan/pull/3/commits/eb4ff132b95a7fa5c54e3541d3b7675aa1c3d7b4

Fix version 3 ported to ckan 2.6.0 card:
https://trello.com/c/1tDpPqip/449-make-ckan-260-version-of-upload-file-tmp-file-space-leak-fix
